uc-crux-llvm/src/UCCrux/LLVM/Classify.hs

```diff
@@ -0,0 +1,643 @@
+                     Nothing -> subAnns
+
+             argName :: Ctx.Index argTypes tp -> String
+             argName idx = funCtx ^. argumentNames . ixF' idx . to getConst . to (maybe "<top>" Text.unpack)
+
+             elemsFromOffset ::
+               What4.ConcreteVal (What4.BaseBVType w) ->
+               PartTypeRepr m ft ->
+               Int
+             elemsFromOffset offset partType =
+               let dl = modCtx ^. dataLayout
+                   pointedTo = asFullType (funCtx ^. moduleTypes) partType
+                   typeSize = bytesToInteger (memTypeSize dl (toMemType pointedTo))
+                in 1 + fromIntegral (asUnsigned (What4.fromConcreteBV offset) `div` fromIntegral typeSize)
+
+             oneArgConstraint :: Ctx.Index argTypes inTy -> Cursor m inTy atTy -> Constraint m atTy -> [NewConstraint m argTypes]
+             oneArgConstraint idx cursor constraint =
+               [NewConstraint (SomeInSelector (SelectArgument idx cursor)) constraint]
+
+             oneArgShapeConstraint :: Ctx.Index argTypes inTy -> Cursor m inTy atTy -> ShapeConstraint m atTy -> [NewConstraint m argTypes]
+             oneArgShapeConstraint idx cursor shapeConstraint =
+               [NewShapeConstraint (SomeInSelector (SelectArgument idx cursor)) shapeConstraint]
+
+             unclass =
+               do
+                 liftIO $ (appCtx ^. log) Hi ("Couldn't classify error." :: Text)
+                 pure $ ExUncertain (UUnclassified (ppBB badBehavior))
+
+             unfixed :: Unfixed -> IO (Explanation m arch argTypes)
+             unfixed tag =
+               do
+                 (appCtx ^. log) Hi "Prognosis: Fixable, but the fix is not yet implemented."
+                 pure $ ExUncertain (UUnfixed tag)
+
+             unfixable :: Unfixable -> IO (Explanation m arch argTypes)
+             unfixable tag =
+               do
+                 (appCtx ^. log) Hi "Prognosis: Don't know how to fix this."
+                 pure $ ExUncertain (UUnfixable tag)
+          in case badBehavior of
+               LLVMErrors.BBUndefinedBehavior
+                 (LLVMErrors.WriteBadAlignment ptr alignment) ->
+                   case getPtrOffsetAnn (Crucible.unRV ptr) of
+                     Just (Some (TypedSelector ftRepr (SomeInSelector (SelectArgument idx cursor)))) ->
+                       do
+                         let tag = ArgWriteBadAlignment
+                         liftIO $
+                           (appCtx ^. log) Hi $
+                             Text.unwords
+                               [ "Diagnosis: ",
+                                 diagnose tag,
+                                 "#" <> Text.pack (show (Ctx.indexVal idx)),
+                                 "at",
+                                 Text.pack (show (ppCursor (argName idx) cursor)),
+                                 "(" <> Text.pack (show (LLVMPtr.ppPtr (Crucible.unRV ptr))) <> ")"
+                               ]
```

Comment:
Are these bodies all basically duplicated? It is hard to eyeball.  Is there something that can be shared here?

Reply:
If you're referring to the bodies of each branch of the overall `case`, they're not duplicated - I definitely understand how it's hard to eyeball, but this is where a lot of the most important logic of the algorithm is implemented, namely the decisions that classify each error as either a true or false positive, and if it's a false positive, try to deduce a new precondition that could be added to avoid it.

If you're referring specifically to the logging code, then they're *mostly* duplicated, but that's essentially incidental - A lot of the currently-implemented heuristics work with single pointers that appear directly in an argument, but as this function grows there will be others that deal with globals, some that deal with integers in arguments, etc. 

Since this code uses two variables that are from pattern matches (`idx` and `ptr`) I didn't un-inline it and pass them as arguments, but I definitely can.

Reply:
I was just looking at the logging code - I did see that the surrounding things are different

Reply:
After looking over this code again today, there's actually even less duplication that could be pulled out than I remember... While the example highlighted above is a "missing precondition" heuristic and there are a few of those, there are also cases where we just can't do anything reasonable (e.g. addition of a complex symbolic integer to a complex pointer expression), cases where we know what to do but haven't done it yet, etc. and all of those call different functions inside their `unwords` expression.

So if we were to factor this out, we'd have a lot of logging methods:

- One for heuristics dealing with a single pointer in an argument and deciding it's a missing precondition
- One for heuristics dealing with a single pointer in an argument and deciding it's a true positive
- One for heuristics dealing with a single pointer in an argument ...
- One for heuristics dealing with a single integer in an argument ...
- ...

So I'd propose we leave it as-is for now. As the code grows and we're able to identify new abstractions, we may be able to conceptually deduplicate some of these, but with logging code it's hard to be both general and useful. Thoughts?

Reply:
I don't think it has to change, but even a simple outlining of the shell of that message might help make it easier to read:
```
printDiagnosis diag idx cursor ptr =
                             Text.unwords
                               [ "Diagnosis: ",
                                 diag,
                                 "#" <> Text.pack (show (Ctx.indexVal idx)),
                                 "at",
                                 Text.pack (show (ppCursor (argName idx) cursor)),
                                 "(" <> Text.pack (show (LLVMPtr.ppPtr (Crucible.unRV ptr))) <> ")"
                               ]
```
You could then call `printDiagnosis (diagnose tag) idx cursor ptr` in a few places (and `printDiagnosis (ppUnfixed tag) ...` in others)